src/rest-server/src/utils/frameworkConverter.js

```diff
@@ -379,7 +466,143 @@ const convertTaskDetail = async (
+    ),
+    currentAttemptLaunchedTime:
+      new Date(attemptStatus.runTime || attemptStatus.startTime).getTime() ||
+      null,
+    currentAttemptCompletedTime:
+      new Date(attemptStatus.completionTime).getTime() || null,
+    containerId: attemptStatus.podUID,
+    containerIp: attemptStatus.podHostIP,
+    containerNodeName: attemptStatus.podNodeName,
+    // Job level info
+    containerPorts,
+    containerGpus,
+    containerLog: `http://${attemptStatus.podHostIP}:${process.env.LOG_MANAGER_PORT}/log-manager/tail/${userName}/${logPathInfix}/${taskRoleName}/${attemptStatus.podUID}/`,
+    containerExitCode: completionStatus ? completionStatus.code : null,
+    containerExitSpec: completionStatus
+      ? generateExitSpec(completionStatus.code)
+      : generateExitSpec(null),
+    containerExitDiagnostics: exitDiagnostics
+      ? exitDiagnostics.diagnosticsSummary
+      : null,
+    ...(launcherConfig.enabledHived && {
+      hived: {
+        affinityGroupName,
+        lazyPreempted: null,
+        lazyPreemptionStatus: null,
+      },
+    }),
+  };
+};
+
+const convertToTaskDetail = async (
+  attemptFramework,
+  taskRoleName,
+  taskStatus,
+  taskHistories,
+) => {
+  const lastTaskAttemptStatus = taskStatus.attemptStatus;
+  const completionStatus = lastTaskAttemptStatus.completionStatus;
+  const userName = attemptFramework.metadata.labels.userName;
+  const jobName = attemptFramework.metadata.annotations.jobName;
+
+  const taskDetail = {
+    // framework level information
+    frameworkName: userName + '~' + jobName,
+    jobAttemptId: attemptFramework.status.attemptStatus.id,
+    // task role level information
+    taskRoleName: taskRoleName,
+    // task level information
+    taskIndex: taskStatus.index,
+    taskUid: taskStatus.instanceUID,
+    taskState: convertState(
+      taskStatus.state,
+      completionStatus ? completionStatus.code : null,
+    ),
+    retries: taskStatus.retryPolicyStatus.totalRetriedCount,
+    accountableRetries: taskStatus.retryPolicyStatus.accountableRetriedCount,
+    createdTime: new Date(taskStatus.startTime).getTime(),
+    completedTime: new Date(taskStatus.completionTime).getTime(),
+    // task attempt level information
+    attempts: [],
+  };
+
+  const logPathInfix = attemptFramework.metadata.annotations.logPathInfix
+    ? attemptFramework.metadata.annotations.logPathInfix
+    : jobName;
+
+  const ports = attemptFramework.spec.taskRoles.find(
+    (taskRoleSpec) => taskRoleSpec.name === taskRoleName,
+  ).task.pod.metadata.annotations['rest-server/port-scheduling-spec'];
+
+  // fill task attempt level information
+  // last task attempt
+  taskDetail.attempts.push(
+    await convertTaskAttempt(
+      logPathInfix,
+      ports,
+      taskStatus,
+      taskRoleName,
+      userName,
+      lastTaskAttemptStatus,
+    ),
+  );
+
+  // history task attempts
+  for (const taskHistory of taskHistories) {
+    taskDetail.attempts.push(
```

Comment:
Sort the task attempts list by attemptid?

Reply:
> Sort the task attempts list by attemptid?

Fixed by keeping the original history attempts order and append the latest attempt to the end.

Reply:
Seems still out of order?
[https://int.org/rest-server/api/v2/jobs/yqwang~taskhistory-elastic-3/attempts/1/taskRoles/ra/taskIndex/2/attempts](https://int.org/rest-server/api/v2/jobs/yqwang~taskhistory-elastic-3/attempts/1/taskRoles/ra/taskIndex/2/attempts)

---
In reply to: [504547504](https://github.com/microsoft/pai/pull/4958#discussion_r504547504) [](ancestors = 504547504)

Reply:
May be sorted in decreasing order is better. So Mingliang will not need to sort it again, and user does not think the array always contains all attempts(some may miss in DB). 

---
In reply to: [506148119](https://github.com/microsoft/pai/pull/4958#discussion_r506148119) [](ancestors = 506148119,504547504)